File: db4on/core/src/core/com/db4o/inside/classindex/BTreeClassIndexStrategy.cs
namespace com.db4o.inside.classindex
{
	/// <exclude></exclude>
	public class BTreeClassIndexStrategy : com.db4o.inside.classindex.AbstractClassIndexStrategy
	{
		private com.db4o.inside.btree.BTree _btreeIndex;

		public BTreeClassIndexStrategy(com.db4o.YapClass yapClass) : base(yapClass)
		{
		}

		public virtual com.db4o.inside.btree.BTree Btree()
		{
			return _btreeIndex;
		}

		public override int EntryCount(com.db4o.Transaction ta)
		{
			return _btreeIndex != null ? _btreeIndex.Size(ta) : 0;
		}

		public override void Initialize(com.db4o.YapStream stream)
		{
			CreateBTreeIndex(stream, 0);
		}

		public override void Purge()
		{
		}

		public override void Read(com.db4o.YapStream stream, int indexID)
		{
			ReadBTreeIndex(stream, indexID);
		}

		public override int Write(com.db4o.Transaction trans)
		{
			if (_btreeIndex == null)
			{
				return 0;
			}
			_btreeIndex.Write(trans);
			return _btreeIndex.GetID();
		}

		public override void TraverseAll(com.db4o.Transaction ta, com.db4o.foundation.Visitor4
			 command)
		{
			if (_btreeIndex != null)
			{
				_btreeIndex.TraverseKeys(ta, command);
			}
		}

		private void CreateBTreeIndex(com.db4o.YapStream stream, int btreeID)
		{
			if (stream.IsClient())
			{
				return;
			}
			_btreeIndex = ((com.db4o.YapFile)stream).CreateBTreeClassIndex(btreeID);
			_btreeIndex.SetRemoveListener(new _AnonymousInnerClass61(this, stream));
		}

		private sealed class _AnonymousInnerClass61 : com.db4o.foundation.Visitor4
		{
			public _AnonymousInnerClass61(BTreeClassIndexStrategy _enclosing, com.db4o.YapStream
				 stream)
			{
				this._enclosing = _enclosing;
				this.stream = stream;
			}

			public void Visit(object obj)
			{
				int id = ((int)obj);
				com.db4o.YapObject yo = stream.GetYapObject(id);
				if (yo != null)
				{
					stream.RemoveReference(yo);
				}
			}

			private readonly BTreeClassIndexStrategy _enclosing;

			private readonly com.db4o.YapStream stream;
		}

		private void ReadBTreeIndex(com.db4o.YapStream stream, int indexId)
		{
			if (!stream.IsClient() && _btreeIndex == null)
			{
				CreateBTreeIndex(stream, indexId);
			}
		}

		protected override void InternalAdd(com.db4o.Transaction trans, int id)
		{
			_btreeIndex.Add(trans, id);
		}

		protected override void InternalRemove(com.db4o.Transaction ta, int id)
		{
			_btreeIndex.Remove(ta, id);
		}

		public override void DontDelete(com.db4o.Transaction transaction, int id)
		{
		}

		public override void DefragReference(com.db4o.YapClass yapClass, com.db4o.YapReader
			 source, com.db4o.YapReader target, com.db4o.IDMapping mapping, int classIndexID
			)
		{
			source.ReadInt();
			int newID = -classIndexID;
			target.WriteInt(newID);
		}

		public override int Id()
		{
			return _btreeIndex.GetID();
		}

		public override com.db4o.foundation.Iterator4 AllSlotIDs(com.db4o.Transaction trans
			)
		{
			return _btreeIndex.AllNodeIds(trans);
		}

		public override void DefragIndex(com.db4o.YapReader source, com.db4o.YapReader target
			, com.db4o.IDMapping mapping)
		{
			_btreeIndex.DefragIndex(source, target, mapping);
		}
	}
}

File: db4oj/core/src/com/db4o/internal/classindex/BTreeClassIndexStrategy.java
/* Copyright (C) 2004 - 2006  db4objects Inc.  http://www.db4o.com */

package com.db4o.internal.classindex;

import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.btree.*;

/**
 * @exclude
 */
public class BTreeClassIndexStrategy extends AbstractClassIndexStrategy {
	
	private BTree _btreeIndex;
	
	public BTreeClassIndexStrategy(ClassMetadata yapClass) {
		super(yapClass);
	}	
	
	public BTree btree() {
		return _btreeIndex;
	}

	public int entryCount(Transaction ta) {
		return _btreeIndex != null
			? _btreeIndex.size(ta)
			: 0;
	}

	public void initialize(ObjectContainerBase stream) {
		createBTreeIndex(stream, 0);
	}

	public void purge() {
	}

	public void read(ObjectContainerBase stream, int indexID) {
		readBTreeIndex(stream, indexID);
	}

	public int write(Transaction trans) {
		if (_btreeIndex == null){
            return 0;
        } 
        _btreeIndex.write(trans);
        return _btreeIndex.getID();
	}
	
	public void traverseAll(Transaction ta,Visitor4 command) {
		// better alternatives for this null check? (has been moved as is from YapFile)
		if(_btreeIndex!=null) {
			_btreeIndex.traverseKeys(ta,command);
		}
	}

	private void createBTreeIndex(final ObjectContainerBase stream, int btreeID){
        if (stream.isClient()) {
        	return;
        }
        _btreeIndex = ((LocalObjectContainer)stream).createBTreeClassIndex(btreeID);
        _btreeIndex.setRemoveListener(new Visitor4() {
            public void visit(Object obj) {
            	removeId((TransactionContext)obj);
            }
        });
    }
	
    private void removeId(TransactionContext context) {
    	ReferenceSystem referenceSystem = context._transaction.referenceSystem();
        ObjectReference reference = referenceSystem.referenceForId((Integer)context._object);
        if(reference != null){
            referenceSystem.removeReference(reference);
        }
    }

	private void readBTreeIndex(ObjectContainerBase stream, int indexId) {
		if(! stream.isClient() && _btreeIndex == null){
            createBTreeIndex(stream, indexId);
		}
	}

	protected void internalAdd(Transaction trans, int id) {
		_btreeIndex.add(trans, new Integer(id));
	}

	protected void internalRemove(Transaction ta, int id) {
		_btreeIndex.remove(ta, new Integer(id));
	}

	public void dontDelete(Transaction transaction, int id) {
	}
	
	public void defragReference(ClassMetadata classMetadata, DefragmentContextImpl context,int classIndexID) {
		int newID = -classIndexID;
		context.writeInt(newID);
	}
	
	public int id() {
		return _btreeIndex.getID();
	}

	public Iterator4 allSlotIDs(Transaction trans) {
        return _btreeIndex.allNodeIds(trans);
	}

	public void defragIndex(DefragmentContextImpl context) {
		_btreeIndex.defragIndex(context);
	}
	
	public static BTree btree(ClassMetadata clazz) {
		ClassIndexStrategy index = clazz.index();
		if(! (index instanceof BTreeClassIndexStrategy)){
			throw new IllegalStateException();
		}
		return ((BTreeClassIndexStrategy)index).btree();
	}
	
	public static Iterator4 iterate(ClassMetadata clazz, Transaction trans) {
		return btree(clazz).asRange(trans).keys();
	}
	
	
	
}
